python/gigl/distributed/utils/neighborloader.py

```diff
@@ -27,26 +28,49 @@ def patch_fanout_for_sampling(
     https://github.com/alibaba/graphlearn-for-pytorch/blob/26fe3d4e050b081bc51a79dc9547f244f5d314da/graphlearn_torch/python/distributed/dist_neighbor_sampler.py#L317-L318
 
     Args:
-        edge_types (list[EdgeType]): List of all edge types in the graph.
+        edge_types (Optional[list[EdgeType]]): List of all edge types in the graph, is None for homogeneous datasets
         num_neighbors (dict[EdgeType, list[int]]): Specified fanout by the user
     Returns:
-        dict[EdgeType, list[int]]: Modified fanout that is approariate for sampling.
+        Union[list[int], dict[EdgeType, list[int]]]: Modified fanout that is appropriate for sampling. Is a list[int]
+            if the dataset is homogeneous, otherwise is dict[EdgeType, list[int]]
     """
+    if edge_types is None:
+        if isinstance(num_neighbors, abc.Mapping):
+            raise ValueError(
+                "When dataset is homogeneous, the num_neighbors field cannot be a dictionary."
+            )
+        return num_neighbors
     if isinstance(num_neighbors, list):
         original_fanout = num_neighbors
+        should_broadcast_fanout = True
         num_neighbors = {}
     else:
+        for edge_type in num_neighbors.keys():
+            if edge_type not in edge_types:
+                raise ValueError(
+                    f"Found edge type {edge_type} in fanout which is not in dataset edge types {edge_types}."
+                )
         original_fanout = next(iter(num_neighbors.values()))
+        should_broadcast_fanout = False
         num_neighbors = deepcopy(num_neighbors)
 
     num_hop = len(original_fanout)
     zero_samples = [0 for _ in range(num_hop)]
     for edge_type in edge_types:
         if is_label_edge_type(edge_type):
             num_neighbors[edge_type] = zero_samples
-        elif edge_type not in num_neighbors:
+        elif should_broadcast_fanout and edge_type not in num_neighbors:
             num_neighbors[edge_type] = original_fanout
-
+        elif not should_broadcast_fanout and edge_type not in num_neighbors:
+            raise ValueError(
+                f"Found non-labeled edge type in dataset {edge_type} which is not in the provided fanout {num_neighbors.keys()}"
```

Comment:
```suggestion
                f"Found non-labeled edge type in dataset {edge_type} which is not in the provided fanout {num_neighbors.keys()}. If fanout is provided as a dict, all edges must be present."
```

Let's say the expectation here?

Also maybe in the doc comment?